[cutlass backend] Add instantiation level for generating configs

henrylhtsang opened the pull request:
Passing through instantiation level to generate more configs.

I do see some C++ compilation error. But running is fine. Using 2222 generates 1k+ configs. 


Differential Revision: D68989194

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @desertfire @chauhang @aakhundov

Chillee commented:
cc: @mlazos

henrylhtsang commented:
Just a heads up, right now the CUTLASS backend is pretty fragile. For example when I use instantiation level = "2222", I am seeing 24 C++ compile error that looks like

```
Output:
cutlass/gemm/collective/builders/sm90_common.inl(145): error: static assertion failed
      static_assert(threads_minor == 0 || (TileSizeMN % threads_minor == 0));
      ^
          detected during:
cutlass/gemm/collective/builders/sm90_gmma_builder.inl
            instantiation of class "cutlass::gemm::collective::CollectiveBuilder<cutlass::arch::Sm90, cutlass::arch::OpClassTensorOp, ElementA, GmemLayoutATag, AlignmentA, ElementB, GmemLayoutBTag, AlignmentB, ElementAccumulator, TileShape_MNK, ClusterShape_MNK, StageCountType, KernelScheduleType, std::enable_if_t<<expression>, void>> [with ElementA=cutlass::bfloat16_t, GmemLayoutATag=cutlass::layout::RowMajor, AlignmentA=8, ElementB=cutlass::bfloat16_t, GmemLayoutBTag=cutlass::layout::ColumnMajor, AlignmentB=8, ElementAccumulator=float, TileShape_MNK=cute::tuple<cute::_128, cute::_16, cute::_64>, ClusterShape_MNK=cute::tuple<cute::_1, cute::_1, cute::_1>, StageCountType=cutlass::gemm::collective::StageCountAutoCarveout<1>, KernelScheduleType=cutlass::gemm::KernelCpAsyncWarpSpecialized]" at line 98 of /tmp/torchinductor_henrylhtsang/tmp1fefkur3/75/c75sitloxfuyyltxsayco3vzzhugcwpqqix4wkhq4mkfuay57mdb.cu
```

The overall compilation is successful. 
```
SingleProcess AUTOTUNE benchmarking takes 547.8341 seconds and 1158.2490 seconds precompiling for 1305 choices
```

henrylhtsang commented:
I also did get an inductor timeout error when using instantiation level 9999

Chillee commented:
Are you using cutlass 3.7? They fixed a bunch of these issues in that release.

henrylhtsang commented:
> Are you using cutlass 3.7? They fixed a bunch of these issues in that release.

Yes I am using internal, which I upgraded to 3.7 earlier.

Chillee commented on torch/_inductor/config.py:
```diff
@@ -1277,6 +1277,16 @@ class cuda:
+    # increasing third digit generates more MMA multipliers,
+    # increasing fourth digit generates more instruction shapes.
+    cutlass_instantiation_level: str = os.environ.get(
+        "CUTLASS_INSTANTIATION_LEVEL", "0"
```
Add a `TORCHINDUCTOR` prefix in front.

Chillee replied:
Also, should it default to 0? Or 121?

henrylhtsang replied:
@Chillee 0 is fine. If it is 0, then it will use the default level (which for most cases is either 121 or 111)

https://github.com/NVIDIA/cutlass/blob/affd1b693dfc121c51118cbc8583dfd308227ca6/python/cutlass_library/manifest.py#L573-L591

henrylhtsang committed:
[cutlass backend] Add instantiation level for generating configs (#146230)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/146230

Differential Revision: D68989194

henrylhtsang commented:
@pytorchbot merge
